### PR TITLE
Add org.artisan_scope.artisan

### PR DESCRIPTION
**Pull request by wvengen**
### Please confirm your submission meets all the criteria

- [x] Please describe the application briefly.
  _Artisan is an open-source software that helps coffee roasters record, analyze, and control roast profiles. This software can automate the creation of roasting metrics to help make decisions that influence the final coffee flavor._

- [x] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read the and followed all the [Submission and licence requirements][reqs].
- [x] I have [built][build] and tested the submission locally.
- [x] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about this submission. **Link:** https://github.com/artisan-roaster-scope/artisan/issues/1602

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

File: org.artisan_scope.artisan.yml [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  - --socket=wayland
+  #- --device=dri # we already have --device=all
+  # USB
+  - --device=all
+  # allow network access for network devices and artisan.plus
+  - --share=network
+  # storing and retrieving artisan.plus credentials
+  - --talk-name=org.freedesktop.secrets
+  # filesystem access needed for autosave
+  - --filesystem=xdg-documents
```
Comment (wvengen) [inline org.artisan_scope.artisan.yml:55]: I tried making it work without xdg-documents access, but couldn't get autosave to work. Some things depend on the version of the xdg desktop portal in use, and in theory even opening a directory should work, but I never got it working. To make this work for various users, I chose the xdg-documents approach, allowing autosave to work in the user's Documents directory, but not expose all files. I've seen this in other flatpaks as well, and hope this line can be removed at some point.

File: org.artisan_scope.artisan.yml [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      - python3 -m compileall $FLATPAK_DEST/{artisanlib,plus,proto,help,uic,misc}
+      # Flatpak expects Qt plugins in lib/plugins/ instead of plugins/
+      - ln -s ../plugins $FLATPAK_DEST/lib/
+      # desktop integration
+      - ln -s /app/artisan.py /app/bin/artisan
+      - mkdir -p $FLATPAK_DEST/share/{metainfo,applications,icons/hicolor/scalable/apps}
+      - install -Dm644 src/artisan.svg $FLATPAK_DEST/share/icons/hicolor/scalable/apps/$FLATPAK_ID.svg
+      - install -Dm644 src/debian/usr/share/metainfo/$FLATPAK_ID.metainfo.xml $FLATPAK_DEST/share/metainfo/
+      - install -Dm644 src/debian/usr/share/applications/$FLATPAK_ID.desktop $FLATPAK_DEST/share/applications/
+      - sed -i "s/^Icon=.*$/Icon=$FLATPAK_ID/" $FLATPAK_DEST/share/applications/$FLATPAK_ID.desktop
```
Comment (wvengen) [inline org.artisan_scope.artisan.yml:94]: I'm not that happy with installation procedure. Upstream only makes fully bundled packages, including Qt and friends, which is not so appropriate for Flatpaks. So after some discussion upstream, we decided to have the installation procedure here in the Flatpak, and gradually move (hopefully most of it) to a more integrated installation approach upstream.

File: TESTING.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+Some packages have built-in tests that can be run.
+
+Note that this is especially important to do when building OpenBLAS, NumPy and/or SciPy from source,
+and upgrading one of them, as they strongly depend on each other, and may not always be compatible
+(e.g. [scipy#19831](https://github.com/scipy/scipy/issues/19831))
+
+The Flatpak does by default not include everything to run these tests, so you'll
+need to build an adapted Flatpak to run them.
+
+## numpy & scipy
```
Comment (wvengen) [inline TESTING.md:11]: As a sidenote: I did extensive testing on NumPy and Scipy, and only got their tests to work if I used the wheels, as much as I wanted to build from source.

File: dep-lib-qhull.yml
Comment (hfiguiere) [inline dep-lib-qhull.yml:1]: this should be in the main manifest instead since it's not autogenearated.

File: dep-lib-phidgets.yml
Comment (hfiguiere) [inline dep-lib-phidgets.yml:1]: this should be in the main manifest instead since it's not autogenearated.

**Comment by hfiguiere** [comment]
bot, build org.artisan_scope.artisan

**Comment by wvengen** [comment]
Did a test with the testing repository, and it seems to function well (just like a local build).

File: org.artisan_scope.artisan.yml (hunk 2 above)
Comment (bbhtt) [reply org.artisan_scope.artisan.yml:118]: Use `desktop-file-edit` instead of sed here to change the icon.

**Comment by bbhtt** [comment]
I'm a bit wary of the extensive cleanup commands. How much are you saving doing them? If the space saving is minimal it's not worth the added complexity there.

File: org.artisan_scope.artisan.yml (hunk 2 above)
Comment (wvengen) [reply org.artisan_scope.artisan.yml]: Thank you, that is much better!

File: org.artisan_scope.artisan.yml [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+sdk: org.kde.Sdk
+runtime-version: '6.7'
+
+base: com.riverbankcomputing.PyQt.BaseApp
+base-version: '6.7'
+cleanup-commands:
+  # uninstall build dependencies
+  - python3 -m pip uninstall -y meson-python cppy pybind11
+  # deduplicate (matplotlib) fonts already in runtime
+  - for f in /usr/share/fonts/dejavu/*.ttf; do find /app -name `basename $f` -execdir ln -sf "$f" \; ; done
```
[the remark this answers is not in the document]
Comment (wvengen) [reply org.artisan_scope.artisan.yml:10]: Saving the space for these standard fonts already present in the filesystem.

File: org.artisan_scope.artisan.yml [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+      # copy all gathered files to the destination
+      - mkdir $FLATPAK_DEST/includes
+      - cp -Rp src/dist/* $FLATPAK_DEST/includes
+      # and also the Python files
+      - cp -Rp src/{artisan.py,artisanlib,plus,proto,help,uic,misc} $FLATPAK_DEST
+      - chmod a+x $FLATPAK_DEST/artisan.py
+      # fix translations not being found by default
+      - mv $FLATPAK_DEST/includes/translations/* $FLATPAK_DEST/translations
+      - rmdir $FLATPAK_DEST/includes/translations
+      - sed -i "s/trans_paths\.append('translations')/trans_paths.append('\/app\/translations')/" $FLATPAK_DEST/artisanlib/main.py
```
[the remark this answers is not in the document]
Comment (wvengen) [reply org.artisan_scope.artisan.yml]: Ah, yes, that's way more robust. Thanks for the suggestion.

**Comment by wvengen** [comment]
> How much are you saving doing them?

The Flatpak is 625 MB with cleanup, 966MB without (with baseapp cleanup and build dependency removal), reported as "Installed" by `flatpak info`.

**Comment by wvengen** [comment]
A bit more background for the cleanup:
- Remove QtWebEngine dictionaries - save 216MB
- Remove Python package tests - save 48MB
- Remove QtWebEngine developer tools & webdriver - save 25MB
- Cleanup unused locale data - save 10MB
- Remove [Yoctupuce libs](https://github.com/yoctopuce/yoctolib_python/issues/11) for other platforms - save 10MB
- Symlink DejaVu fonts - save 5MB

I think the cleaning up of unused locale data is the most error-prone and doesn't add that much, I've removed that.

**Comment by bbhtt** [comment]
bot, build org.artisan_scope.artisan

**Pull request closed without merging**